builder/cloudstack/step_create_template.go

```diff
@@ -0,0 +1,103 @@
+package cloudstack
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/mitchellh/multistep"
+	"github.com/mitchellh/packer/packer"
+	"github.com/xanzy/go-cloudstack/cloudstack"
+)
+
+type stepCreateTemplate struct{}
+
+func (s *stepCreateTemplate) Run(state multistep.StateBag) multistep.StepAction {
+	client := state.Get("client").(*cloudstack.CloudStackClient)
+	config := state.Get("config").(*Config)
+	ui := state.Get("ui").(packer.Ui)
+
+	ui.Say(fmt.Sprintf("Creating template: %s", config.TemplateName))
+
+	// Retrieve the instance ID from the previously saved state.
+	instanceID, ok := state.Get("instance_id").(string)
+	if !ok || instanceID == "" {
+		ui.Error("Could not retrieve instance_id from state!")
+		return multistep.ActionHalt
+	}
+
+	// Create a new parameter struct.
+	p := client.Template.NewCreateTemplateParams(
+		config.TemplateDisplayText,
+		config.TemplateName,
+		config.TemplateOS,
+	)
+
+	// Configure the template according to the supplied config.
+	p.SetIsfeatured(config.TemplateFeatured)
+	p.SetIspublic(config.TemplatePublic)
+	p.SetIsdynamicallyscalable(config.TemplateScalable)
+	p.SetPasswordenabled(config.TemplatePasswordEnabled)
+	p.SetRequireshvm(config.TemplateRequiresHVM)
+
+	if config.Project != "" {
+		p.SetProjectid(config.Project)
+	}
+
+	if config.TemplateTag != "" {
+		p.SetTemplatetag(config.TemplateTag)
+	}
+
+	ui.Message("Retrieving the ROOT volume ID...")
+	volumeID, err := getRootVolumeID(client, instanceID)
+	if err != nil {
+		ui.Error(err.Error())
+		return multistep.ActionHalt
+	}
+
+	// Set the volume ID from which to create the template.
+	p.SetVolumeid(volumeID)
+
+	ui.Message("Creating the new template...")
+	template, err := client.Template.CreateTemplate(p)
+	if err != nil {
+		ui.Error(fmt.Sprintf("Error creating the new template %s: %s", config.TemplateName, err))
+		return multistep.ActionHalt
+	}
+
+	// This is kind of nasty, but it appears to be needed to prevent corrupt templates.
+	// When CloudStack says the template creation is done and you then delete the source
+	// volume shortly after, it seems to corrupt the newly created template. Giving it an
+	// additional 30 seconds to really finish up, seem to prevent that from happening.
```

Comment:
I think a 30s stop would make me wounder what's going on. It's not super important.